Formatter/Pool.php

```diff
@@ -29,20 +31,56 @@ class Pool
     protected $defaultFormatter;
 
     /**
+     * NEXT_MAJOR: use LoggerAwareTrait.
+     *
      * @var LoggerInterface
      */
     protected $logger;
 
     /**
-     * @param LoggerInterface|null $logger
-     * @param string|null          $defaultFormatter
+     * @param string|null $defaultFormatter
      */
-    public function __construct(LoggerInterface $logger = null, $defaultFormatter = null)
+    public function __construct($defaultFormatter = null)
     {
-        $this->logger = $logger;
-
-        // TODO: This should become a required first parameter when the major version is changed
+        $this->logger = new NullLogger();
+        // NEXT_MAJOR: remove this block
+        if (func_num_args() == 2) {
+            list($logger, $defaultFormatter) = func_get_args();
+            $this->logger = $logger;
+            $this->defaultFormatter = $defaultFormatter;
+
+            return;
+        }
+        if (func_num_args() == 1 && $defaultFormatter instanceof LoggerInterface) {
+            $this->logger = $defaultFormatter;
+            @trigger_error(sprintf(
+                'Providing a logger to %s through the constructor is deprecated since 3.x'.
+                ' and we no longer be possible in 4.0'.
+                ' This argument should be provided through the setLogger() method.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
+
+            return;
+        }
         $this->defaultFormatter = $defaultFormatter;
+        // NEXT_MAJOR: make defaultFormatter required
+        if (is_null($defaultFormatter)) {
+            @trigger_error(sprintf(
+                'Not providing the defaultFormatter argument to %s is deprecated since 3.x.'.
+                ' This argument will become mandatory in 4.0.',
+                __METHOD__
+            ), E_USER_DEPRECATED);
+        }
+    }
+
+    /**
+     * NEXT_MAJOR: use Psr\Log\LoggerAwareTrait.
+     *
+     * @param LoggerInterface will be used to report errors
+     */
+    final public function setLogger(LoggerInterface $logger)
```

Comment:
Not sure how this works in the interface/trait, but I would use the exact same signature to avoid breaking BC (though removing final is not a BC breaking operation). As far as I remember the trait exposes a fluent interface. I might be wrong, but if not, this should be fluent as well.